crates/wast/src/binary.rs

```diff
@@ -1108,7 +1228,159 @@ fn find_names<'a>(
     return ret;
 }
 
-impl Names<'_> {
+#[derive(Default)]
+struct ComponentNames<'a> {
+    component: Option<&'a str>,
+    funcs: Vec<(u32, &'a str)>,
+    func_idx: u32,
+    locals: Vec<(u32, Vec<(u32, &'a str)>)>,
+    labels: Vec<(u32, Vec<(u32, &'a str)>)>,
+    components: Vec<(u32, &'a str)>,
+    component_idx: u32,
+    instances: Vec<(u32, &'a str)>,
+    instance_idx: u32,
+    types: Vec<(u32, &'a str)>,
+    type_idx: u32,
+}
+
+fn find_component_names<'a>(
+    component_id: &Option<Id<'a>>,
+    component_name: &Option<NameAnnotation<'a>>,
+    fields: &[ComponentField<'a>],
+) -> ComponentNames<'a> {
+    fn get_name<'a>(id: &Option<Id<'a>>, name: &Option<NameAnnotation<'a>>) -> Option<&'a str> {
+        name.as_ref().map(|n| n.name).or(id.and_then(|id| {
+            if id.is_gensym() {
+                None
+            } else {
+                Some(id.name())
+            }
+        }))
+    }
```

Comment:
I suspect that this is the same implementation as something for modules, so could that be deduplicated?